NuKeeper.Inspection/NuGetApi/VersionChanges.cs

```diff
@@ -43,6 +44,23 @@ private static bool Filter(NuGetVersion v1, NuGetVersion v2, VersionChange allow
                 case VersionChange.Patch:
                     return (v1.Major == v2.Major) && (v1.Minor == v2.Minor);
 
+                case VersionChange.Release:
+                    if (string.IsNullOrEmpty(v1.Release))
+                    {
+                        return (v1.Major == v2.Major) && (v1.Minor == v2.Minor);
+                    }
+                    else
+                    {
+                        if (string.IsNullOrEmpty(v2.Release))
+                        {
+                            return (v1.Major == v2.Major) && (v1.Minor == v2.Minor);
+                        }
+                        else
+                        {
+                            return (v1.Major == v2.Major) && (v1.Minor == v2.Minor) && (v1.Patch == v2.Patch);
+                        }
+                    }
+
```

Comment:
I think this is the key code.  Can you explain why the `Patch` is only checked if both have `Release` tags? Versions that differ only in `Release` would have same Major, Minor and Patch parts?

Also, I would start it with 
``` 
if (string.IsNullOrEmpty(v1.Release)) || string.IsNullOrEmpty(v2.Release))
{
 return ...
}
```

Reply:
Looking at the calling code, this is comparing the `current` version with the available versions collection.


1. If the `current` doesn't have a release tag, it won't matter if the comparing one has one or not, as long as it's higher or the same, it's good to go. Doesn't matter the release here, because, even on the most de-favourable case 1.2.3 against 1.2.3-preview002, the 1st will always win because the second is a preview of the first.

2. If the comparing one doesn't have a release tag, the comparison should be the same as with Patch, because it won't matter.


3. If the comparing one _has_ a release tag (like the current, than we need to compare all three, up to the release one.

Reply:
In the existing code , If the allowedChange is `Minor`, it won't match `1.2.0` to `2.3.0` since the Major is different and only Minor or below can differ. If the allowedChange is `Patch`, it won't match `1.2.0` to `1.3.0` since the Minor is different and only Patch can differ.

You'd expect based on that for allowedChange of `Release` then the Major, Minor and Patch should all match; but it's not the same since release number is optional. So we're saying that `Release` should work as `allowedChange` of `Patch`, unless it's comparing both current and available that both have non-empty release values?